### PR TITLE
Implement book search by parameters

### PR DESCRIPTION
**Pull request by chertiav**
- added a `BookSearchParameters` DTO;
- added a new search endpoint in `BookController`;
- added dependencies for QueryDSL;
- removed unused `DataProcessingException` class.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -66,6 +66,19 @@
             <artifactId>liquibase-maven-plugin</artifactId>
             <version>${liquibase.version}</version>
         </dependency>
+        <dependency>
+            <groupId>com.querydsl</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml]: Why do you need these dependencies?

File: pom.xml [hunk 2]
```diff
@@ -140,17 +153,43 @@
...
+            <plugin>
+                <groupId>org.liquibase</groupId>
+                <artifactId>liquibase-maven-plugin</artifactId>
+                <version>${liquibase.version}</version>
+                <configuration>
+                    <propertyFile>src/main/resources/liquibase.properties</propertyFile>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>com.mysema.maven</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml]: why do you need this plugin?

File: pom.xml [hunk 3]
```diff
@@ -140,17 +153,43 @@
                             <artifactId>mapstruct-processor</artifactId>
                             <version>${org.mapstruct.version}</version>
                         </path>
+                        <path>
+                            <groupId>jakarta.annotation</groupId>
+                            <artifactId>jakarta.annotation-api</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml]: same question here)

File: src/main/java/com/achdev/onlinebookstoreapp/repository/querydsl/QPredicates.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.querydsl.core.types.dsl.BooleanExpression;
+import com.querydsl.core.types.dsl.Expressions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class QPredicates {
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/repository/querydsl/QPredicates.java]: It's nice that you find the other way to implement this task, but let's use JPA Specification to implement searching to learn it

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/achdev/onlinebookstoreapp/repository/querydsl/QPredicates.java (hunk 4 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/repository/querydsl/QPredicates.java]: Ok)

File: src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/TitleSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("title").in(Arrays.stream(params).toArray());
+    }
+
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (okuzan) [inline src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/TitleSpecificationProvider.java]: extract repeated literal to a constant

File: src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/AuthorSpecificationProvider.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("author").in(Arrays.stream(params).toArray());
+    }
+
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (okuzan) [inline src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/AuthorSpecificationProvider.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/TitleSpecificationProvider.java (hunk 5 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/TitleSpecificationProvider.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/AuthorSpecificationProvider.java (hunk 6 above)
Comment (chertiav) [reply src/main/java/com/achdev/onlinebookstoreapp/repository/book/spec/AuthorSpecificationProvider.java]: Ok

File: src/main/java/com/achdev/onlinebookstoreapp/repository/book/BookSpecificationBuilder.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParameters searchParameters) {
+        Specification<Book> bookSpecification = Specification.where(null);
+        if (searchParameters.titles() != null && searchParameters.titles().length > 0) {
+            bookSpecification = bookSpecification
+                    .and(specificationProviderManager.getSpecificationProvider("title")
```
Comment (Elena-Bruyako) [inline src/main/java/com/achdev/onlinebookstoreapp/repository/book/BookSpecificationBuilder.java]: title and author should be constants

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comment

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**